src/cls/queue/cls_queue_src.cc

```diff
@@ -0,0 +1,901 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#include "include/types.h"
+
+#include <errno.h>
+
+#include "objclass/objclass.h"
+#include "cls/queue/cls_queue_types.h"
+#include "cls/queue/cls_queue_ops.h"
+#include "cls/queue/cls_queue_const.h"
+#include "cls/queue/cls_queue.h"
+
+static int get_queue_head_and_size(cls_method_context_t hctx, cls_queue_head& head, uint64_t& head_size)
+{
+  //read head size
+  bufferlist bl_head_size;
+  int ret = cls_cxx_read(hctx, 0, sizeof(uint64_t), &bl_head_size);
```

Comment:
since get_queue_head_and_size() is called by every other op, it would be nice to avoid doing two separate reads for the length and the head, using a strategy similar to cls_queue_list_entries(). for example, given some `block_size`:
* read the first `block_size` bytes into `bufferlist bl`
* decode `uint64_t encoded_size`
* if encoded_size+8 > block_size, read the rest and append to `bl`
* decode `head`

the actual encoded size of `cls_queue_head` will often be much smaller than the maximum `head_size`, so i think it's better to write the `encoded_size` at offset 0, and encode the maximum `head_size` as a member variable inside `cls_queue_head`. this way, we don't have to read the maximum size every time, so we're more likely to get the entire head in the first read

the `encoded_size` can change, so we'd need to keep it updated each time we rewrite the head; i'd suggest a `write_head()` helper function to encapsulate that:
```c++
int write_head(cls_method_context_t hctx, const cls_queue_head& head)
{
  bufferlist bl_head;
  encode(head, bl_head);
  uint64_t encoded_size = bl_head.length();
  bufferlist bl;
  encode(encoded_size, bl);
  bl.claim_append(bl_head);
  return cls_cxx_write2(hctx, 0, bl.length(), &bl, ...);
}
```

Reply:
Ok